### ISSUE TITLE
`parseQuery` decode the query string which cause invalid URI

### ISSUE THREAD

**Issue by kfrawee** [RDFLib/rdflib#1957]
I have the following ttl file:
```ttl
@prefix : <https://www.example.co/reserved/language#> .

<https://www.example.co/reserved/root> :_id "01G39WKRH76BGY5D3SKDHJP2SX" ;
    :transcript%20data [ :_id "01G39WKRH7JYRX78X7FG4RCNYF" ;
            :_key "transcript%20data" ;
            :value "value" ;
            :value_id "01G39WKRH7PVK1DXQHWT08DZA8" ] .
```

And I have the following query:
```python
q = sparql = """
PREFIX : <https://www.example.co/reserved/language#>

    SELECT  ?o 
    WHERE { ?s :transcript%20data/:value ?o . }
""" 
```
While trying to query the graph I got from the ttl file I got the following error:
`https://www.example.co/reserved/language#transcript data does not look like a valid URI, trying to serialize this will break.`

As you see, `parseQuery` has decoded the "%20" to a space " " which cases invalid URI.

I tested the query on different SPARQL engines and it is valid and works as expected.

So, what do you advise?

_I am using rdflib Version: 6.1.1 on macOS Monterey 12.4_

**Comment by kfrawee** [comment]
@aucampia

**Comment by kfrawee** [comment]
@aucampia @gjhiggins

**Comment by aucampia** [comment]
@kfrawee will look when I have time

**Comment by gjhiggins** [comment]
> I have the following ttl file:
...
> So, what do you advise? to make the query valid and get the required results?

Looks like an issue with the SPARQL parser so it might be some time before it's fixed. In the interim, you can work around the problem by replacing `%20` with `_`

**Comment by gjhiggins** [comment]
fwiw:
1. I *think* the causal regexp is https://github.com/RDFLib/rdflib/blob/6ed2ef48ed38679bcdafe7cae250a2ef4b315e7b/rdflib/plugins/sparql/parser.py#L221
2. dunno if it's of any practical use in this instance but it may be worth taking note of / comparing with the example SPARQL 1.1 parser in the [pyastbuilder repos](https://github.com/TNO/pyastbuilder/blob/f202b1cd2bcc7cf5c5ad9ef5402d5ebea490a496/pyastbuilder/parsertools/parsers/sparqlparser.py#L714)

**Comment by gjhiggins** [comment]
Ah, I was close but not close enough.

I'm still exploring the ramifications but if you're prepared to make a small edit to https://github.com/RDFLib/rdflib/blob/6ed2ef48ed38679bcdafe7cae250a2ef4b315e7b/rdflib/plugins/sparql/parser.py#L230 there is a temporary workaround that would enable you to work with the original data:

```diff
 def _hexExpand(match):
-    return chr(int(match.group(0)[1:], 16))
+    if match.group(0) == "%20":
+        return match.group(0)
+    else:
+        return chr(int(match.group(0)[1:], 16))
```

**Comment by kfrawee** [comment]
@gjhiggins
> `if match.group(0) == "%20":`

It is not specific for "%20". I am using urllib.parse.quote to URL encode the ttl and quires. So, the same behavior happens to the rest of the characters.

**Comment by gjhiggins** [comment]
> It is not specific for "%20" ... the same behavior happens to the rest of the characters.

Yes, indeed so. I addressed the entire set of percent encoded reserved characters in the PR https://github.com/RDFLib/rdflib/pull/1959 I submitted.
